apps/cowswap-frontend/src/modules/tokensList/hooks/useRecentTokens.ts

```diff
@@ -0,0 +1,122 @@
+import { useAtomValue, useSetAtom } from 'jotai'
+import { useCallback, useEffect, useMemo } from 'react'
+
+import { TokenWithLogo } from '@cowprotocol/common-const'
+
+import {
+  RECENT_TOKENS_LIMIT,
+  buildFavoriteTokenKeys,
+  buildNextStoredTokens,
+  buildTokensByKey,
+  getStoredTokenKey,
+  hydrateStoredToken,
+  type StoredRecentTokensByChain,
+} from './recentTokensStorage'
+
+import { recentTokensAtom } from '../state/recentTokensAtom'
+import { getTokenUniqueKey } from '../utils/tokenKey'
+
+interface UseRecentTokensParams {
+  allTokens: TokenWithLogo[]
+  favoriteTokens: TokenWithLogo[]
+  activeChainId?: number
+  maxItems?: number
+}
+
+export interface RecentTokensState {
+  recentTokens: TokenWithLogo[]
+  addRecentToken(token: TokenWithLogo): void
+  clearRecentTokens(): void
+}
+
+export function useRecentTokens({
+  allTokens,
+  favoriteTokens,
+  activeChainId,
+  maxItems = RECENT_TOKENS_LIMIT,
+}: UseRecentTokensParams): RecentTokensState {
+  const storedTokensByChain = useAtomValue(recentTokensAtom)
+  const setStoredTokensByChain = useSetAtom(recentTokensAtom)
+
+  const tokensByKey = useMemo(() => buildTokensByKey(allTokens), [allTokens])
+  const favoriteKeys = useMemo(() => buildFavoriteTokenKeys(favoriteTokens), [favoriteTokens])
+
+  // Filter out favorite tokens from stored tokens
+  useEffect(() => {
```

Comment:
As I understand, this hook removes favorite tokens from recent tokens.
But you also filter them out below:

```
if (seenKeys.has(key) || favoriteKeys.has(key)) {
...
```

Looks like this `useEffect` is redundant

Reply:
Addressed:

> I removed that useEffect and now only filter favorites at render time in the useMemo.